In [163]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from env import get_connection

pd.set_option('display.max_columns', None)

import warnings
warnings.filterwarnings('ignore')

In [3]:
url = get_connection('curriculum_logs')

In [116]:
# querying and pulling necessary data
df = pd.read_sql('''
            SELECT * FROM logs as l
            JOIN cohorts as c
            WHERE l.cohort_id = c.id;
            ''', url)

In [117]:
df

date      time  \
0       2018-01-26  09:55:03   
1       2018-01-26  09:56:02   
2       2018-01-26  09:56:05   
3       2018-01-26  09:56:06   
4       2018-01-26  09:56:24   
...            ...       ...   
847325  2021-04-21  16:41:51   
847326  2021-04-21  16:42:02   
847327  2021-04-21  16:42:09   
847328  2021-04-21  16:44:37   
847329  2021-04-21  16:44:39   

                                                    path  user_id  cohort_id  \
0                                                      /        1        8.0   
1                                                java-ii        1        8.0   
2                    java-ii/object-oriented-programming        1        8.0   
3                     slides/object_oriented_programming        1        8.0   
4                              javascript-i/conditionals        2       22.0   
...                                                  ...      ...        ...   
847325                              jquery/personal-site       64       28.0   
847326                                 jquery/mapbox-api       64       28.0   
847327                           jquery/ajax/weather-map       64       28.0   
847328  anomaly-detection/discrete-probabilistic-methods      744       28.0   
847329                                 jquery/mapbox-api       64       28.0   

                   ip  id     name     slack  start_date    end_date  \
0        97.105.19.61   8  Hampton  #hampton  2015-09-22  2016-02-06   
1        97.105.19.61   8  Hampton  #hampton  2015-09-22  2016-02-06   
2        97.105.19.61   8  Hampton  #hampton  2015-09-22  2016-02-06   
3        97.105.19.61   8  Hampton  #hampton  2015-09-22  2016-02-06   
4        97.105.19.61  22    Teddy    #teddy  2018-01-08  2018-05-17   
...               ...  ..      ...       ...         ...         ...   
847325  71.150.217.33  28    Staff         #  2014-02-04  2014-02-04   
847326  71.150.217.33  28    Staff         #  2014-02-04  2014-02-04   
847327  71.150.217.33  28    Staff         #  2014-02-04  2014-02-04   
847328  24.160.137.86  28    Staff         #  2014-02-04  2014-02-04   
847329  71.150.217.33  28    Staff         #  2014-02-04  2014-02-04   

                 created_at           updated_at deleted_at  program_id  
0       2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
1       2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
2       2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
3       2016-06-14 19:52:26  2016-06-14 19:52:26       None           1  
4       2018-01-08 13:59:10  2018-01-08 13:59:10       None           2  
...                     ...                  ...        ...         ...  
847325  2018-12-06 17:04:19  2018-12-06 17:04:19       None           2  
847326  2018-12-06 17:04:19  2018-12-06 17:04:19       None           2  
847327  2018-12-06 17:04:19  2018-12-06 17:04:19       None           2  
847328  2018-12-06 17:04:19  2018-12-06 17:04:19       None           2  
847329  2018-12-06 17:04:19  2018-12-06 17:04:19       None           2  

[847330 rows x 15 columns]

In [118]:
# removing program id 4 
df = df[df['program_id']!=4]

In [119]:
# dropping unnecessary coluns
df.drop(columns=['slack', 'deleted_at','created_at','updated_at'], inplace=True)

/var/folders/mp/b9xckkh51ds5b9r44vtn7w7w0000gn/T/ipykernel_61647/3884105429.py:1: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  df.drop(columns=['slack', 'deleted_at','created_at','updated_at'], inplace=True)


##  Is there a cohort that referred to a lesson significantly more than other cohorts seemed to gloss over?

In [120]:
# removing values with only the slash
df = df[df['path']!='/']

In [122]:
# counting the number of times each path value appears
df['count'] = df.groupby('path')['path'].transform('count')

In [147]:
# splitting data into 3 data frames, 1 for each program
cohort_path_1 = pd.DataFrame(df['path'].drop_duplicates())
cohort_path_2 = pd.DataFrame(df['path'].drop_duplicates())
cohort_path_3 = pd.DataFrame(df['path'].drop_duplicates())

In [148]:
# pulling cohort names for each program 
prgm_1 = df[df['program_id']==1]['name'].unique()
prgm_2 = df[df['program_id']==2]['name'].unique()
prgm_3 = df[df['program_id']==3]['name'].unique()

In [149]:
# filling data frames with information for each path
for name in prgm_1:
    cohort_path_1 = pd.merge(cohort_path_1, pd.DataFrame({f'{name}':df[df['name']==f'{name}'].value_counts('path')}), 
         how='left',on='path')
    
    cohort_path_1[f'{name}'] = cohort_path_1[f'{name}'].fillna(0)

for name in prgm_2:
    cohort_path_2 = pd.merge(cohort_path_2, pd.DataFrame({f'{name}':df[df['name']==f'{name}'].value_counts('path')}), 
         how='left',on='path')
    
    cohort_path_2[f'{name}'] = cohort_path_2[f'{name}'].fillna(0)
    
for name in prgm_3:
    cohort_path_3 = pd.merge(cohort_path_3, pd.DataFrame({f'{name}':df[df['name']==f'{name}'].value_counts('path')}), 
         how='left',on='path')
    
    cohort_path_3[f'{name}'] = cohort_path_3[f'{name}'].fillna(0)

In [164]:
# creating new column with the difference of the most times a lesson was accessed vs the least
cohort_path_1['diff'] = cohort_path_1.max(axis=1) - cohort_path_1.min(axis=1)
cohort_path_2['diff'] = cohort_path_2.max(axis=1) - cohort_path_2.min(axis=1)
cohort_path_3['diff'] = cohort_path_3.max(axis=1) - cohort_path_3.min(axis=1)

In [183]:
# determing highest differences in number of times a lessone was accessed
cohort_path_1[cohort_path_1['diff']>200]

path  Hampton  Arches  Quincy  Kings  Lassen  Glacier  Denali  \
8     javascript-i     36.0   294.0     6.0   17.0   233.0     11.0     0.0   
15          spring     43.0   192.0     8.0    3.0   222.0      2.0     0.0   
20      index.html      1.0     1.0     6.0   84.0   877.0      0.0     0.0   
34        html-css     21.0   215.0    12.0   10.0   174.0     21.0     0.0   
84   javascript-ii     31.0   204.0     7.0   10.0   107.0      6.0     0.0   
131       java-iii     57.0   153.0     3.0    8.0   224.0      0.0     0.0   

     Joshua  Olympic  Badlands   Ike  Franklin   diff  
8       1.0    128.0       1.0   7.0       2.0  294.0  
15      0.0     24.0       0.0   5.0       2.0  222.0  
20      8.0     28.0       0.0   6.0       0.0  877.0  
34      5.0     69.0       2.0  12.0       1.0  215.0  
84      0.0     58.0       0.0   4.0       2.0  204.0  
131     1.0     25.0       0.0   5.0       3.0  224.0

**For program id 1 the cohorts that accessed lessons more than other cohorts was the Lassen cohort**

In [179]:
cohort_path_2[cohort_path_2['diff']>1200]

path  Teddy  Sequoia  Niagara  Pinnacles  Mammoth  \
0                     java-ii  509.0    129.0     17.0       50.0     10.0   
8                javascript-i  615.0    202.0     23.0       69.0      6.0   
15                     spring  744.0    364.0     65.0       56.0     12.0   
16                   appendix  505.0    170.0     20.0       25.0      9.0   
34                   html-css  241.0    168.0      6.0       56.0      3.0   
131                  java-iii  732.0    105.0     29.0       45.0     13.0   
610                       toc   83.0     36.0      8.0       12.0      1.0   
617  search/search_index.json  103.0     45.0      7.0        9.0      1.0   

     Ulysses  Voyageurs  Wrangell  Xanadu  Yosemite   Staff    Zion  \
0      425.0      756.0     325.0   502.0     274.0  1260.0   624.0   
8      526.0      884.0     560.0   722.0     452.0  1817.0   897.0   
15     409.0      650.0     293.0   454.0     381.0  1403.0   672.0   
16     176.0      394.0     209.0   325.0     144.0  1288.0   296.0   
34     562.0      528.0     411.0   597.0     384.0  1284.0   675.0   
131    412.0      770.0     337.0   497.0     351.0  1393.0   753.0   
610     89.0      272.0     991.0   443.0     697.0   911.0  1465.0   
617    142.0      328.0     504.0   577.0     361.0  1349.0   700.0   

     Andromeda  Betelgeuse   Ceres  Deimos  Europa  Fortuna    Apex  Ganymede  \
0        398.0       481.0   681.0   511.0   375.0    657.0   666.0     641.0   
8        516.0       717.0  1003.0   896.0   506.0    785.0   669.0     760.0   
15       487.0       299.0   562.0   533.0   324.0    636.0   600.0     508.0   
16       161.0       188.0   411.0   343.0   243.0    356.0   450.0     561.0   
34       375.0       457.0   766.0   773.0   353.0    585.0   738.0     504.0   
131      433.0       472.0   682.0   523.0   451.0    786.0   663.0     651.0   
610      638.0       610.0   911.0   750.0   964.0   1293.0   810.0     822.0   
617      318.0       761.0  1380.0   746.0   742.0   1020.0  1497.0    1095.0   

     Hyperion   Bash  Jupiter  Kalypso   Luna  Marco  Neptune  Oberon    diff  
0       688.0  268.0    755.0    417.0  255.0   40.0      5.0     0.0  1260.0  
8       910.0  374.0    926.0    631.0  676.0  907.0    272.0   136.0  1811.0  
15      526.0  311.0    621.0    339.0   74.0   43.0      8.0     2.0  1401.0  
16      400.0  174.0    421.0    180.0   92.0   82.0     15.0    10.0  1279.0  
34      439.0  265.0    784.0    495.0  307.0  479.0    313.0    21.0  1281.0  
131     701.0  324.0    795.0    456.0  259.0   40.0      9.0     1.0  1392.0  
610    1022.0  404.0   1866.0    495.0  373.0  910.0    451.0   101.0  1865.0  
617     670.0  660.0    998.0    467.0  405.0  241.0     77.0     9.0  1496.0

**For program id 2 the cohorts that accessed lessons more than other cohorts it was Voyageurs, Zion, Fortuna, Apex, Hyperion, and Jupiter**

In [181]:
cohort_path_3[cohort_path_3['diff']>600]

path  Bayes  Curie  Darden  Easley  \
1391  1-fundamentals/modern-data-scientist.jpg  650.0  467.0   470.0     7.0   
1406  1-fundamentals/1.1-intro-to-data-science  640.0  461.0   460.0     8.0   
1407      1-fundamentals/AI-ML-DL-timeline.jpg  648.0  465.0   470.0     7.0   
1766    fundamentals/modern-data-scientist.jpg   24.0  100.0   308.0   379.0   
1781                        sql/mysql-overview   27.0  125.0   774.0   295.0   
1795                   classification/overview   35.0  136.0  1109.0   445.0   
2003  classification/scale_features_or_not.svg   28.0   96.0   943.0   463.0   

      Florence    diff  
1391      61.0   643.0  
1406      64.0   632.0  
1407      61.0   641.0  
1766     627.0   603.0  
1781     203.0   747.0  
1795      60.0  1074.0  
2003      60.0   915.0

**For program id 3 the cohort that accessed the lessons more than other cohorts it was Darden**

##  Is there any suspicious activity, such as users/machines/etc accessing the curriculum who shouldn’t be? Does it appear that any web-scraping is happening? Are there any suspicious IP addresses?

In [256]:
# determining how many times each ip address shows up
ip_count = pd.DataFrame({'ip_count':df['ip'].value_counts()})

In [251]:
# adding number of times an ip shows up into originial data frame
df = pd.merge(df, ip_count, how='left', left_on='ip', right_on=ip_count.index)

In [204]:
# determing how many ip addresses show up less than 5 times and listing them
sus_ip = pd.DataFrame(df[df['ip_count']<5]['ip'].drop_duplicates())['ip']
sus_ip

187         12.189.101.90
557        70.123.231.211
607       173.173.127.101
808         76.211.179.43
1731      170.248.173.246
               ...       
753157       64.132.4.102
755310     173.174.231.66
763899      13.58.162.201
770857       75.13.87.215
787047      107.77.189.24
Name: ip, Length: 236, dtype: object

**There are 236 ip addresses that were used less than 5 times and should all be considered suspisions id addresses**

## What topics are grads continuing to reference after graduation and into their jobs (for each program)?

In [216]:
# determing which lessons are accessed after a student graduates by program id
cohort_grad_1 = pd.DataFrame({'count':df[(df['program_id']==1) & (df['date']>df['end_date'])].value_counts('path')})

cohort_grad_2 = pd.DataFrame({'count':df[(df['program_id']==2) & (df['date']>df['end_date'])].value_counts('path')})

cohort_grad_3 = pd.DataFrame({'count':df[(df['program_id']==3) & (df['date']>df['end_date'])].value_counts('path')})

In [226]:
# only showing top results
cohort_grad_1[cohort_grad_1['count']>500]

count
path               
index.html     1011
javascript-i    736
html-css        542
spring          501

**For program id 1 the lessons accessed the most after graduation was index.html and javascript-i**

In [220]:
cohort_grad_2[cohort_grad_2['count']>3000]

count
path                           
javascript-i               4229
spring                     3760
search/search_index.json   3562
html-css                   3136
java-iii                   3058

**For program id 2 the lessons accessed the most after graduation was javascript-i and spring**

In [225]:
cohort_grad_3[cohort_grad_3['count']>200]

count
path                                           
search/search_index.json                    493
sql/mysql-overview                          275
classification/overview                     266
classification/scale_features_or_not.svg    219

**For program id 3 the lessons accessed the most after graduation was mysql_overview and classification overview**

## Which lessons are least accessed?

In [259]:
# showing which lessons where accessed the least
df[(df['count']==1) & (df['path'].str.contains('/'))]

date      time  \
203     2018-01-26  13:39:02   
519     2018-01-27  07:49:25   
2623    2018-01-31  15:22:06   
7014    2018-02-09  20:29:05   
15743   2018-03-01  21:07:05   
...            ...       ...   
764977  2021-03-23  14:08:52   
774952  2021-03-31  22:02:18   
786706  2021-04-11  09:17:02   
792346  2021-04-14  22:27:32   
794070  2021-04-15  16:28:50   

                                                     path  user_id  cohort_id  \
203                            javascript/google-maps-api       38       21.0   
519                                              teams/13        1        8.0   
2623                                     javascript/loops       47       21.0   
7014                   content/html-css/search_index.json       80       14.0   
15743               prework/cli/07-editing-files-with-vim      114        8.0   
...                                                   ...      ...        ...   
764977                                       sql/database      949      137.0   
774952                               appendix/3d-plotting      744       28.0   
786706                         classification/explore-old      840      133.0   
792346                                   javascript-i/dom      893       28.0   
794070  appendix/professional-development/post-intervi...        1       28.0   

                    ip   id      name  start_date    end_date  program_id  \
203       97.105.19.61   21   Sequoia  2017-09-27  2018-02-15           2   
519      72.177.226.58    8   Hampton  2015-09-22  2016-02-06           1   
2623      97.105.19.61   21   Sequoia  2017-09-27  2018-02-15           2   
7014     136.50.29.193   14    Lassen  2016-07-18  2016-11-10           1   
15743   173.173.119.75    8   Hampton  2015-09-22  2016-02-06           1   
...                ...  ...       ...         ...         ...         ...   
764977   136.49.16.164  137  Florence  2021-03-15  2021-09-03           3   
774952   24.160.137.86   28     Staff  2014-02-04  2014-02-04           2   
786706  162.222.247.26  133    Easley  2020-12-07  2021-06-08           3   
792346   70.120.31.121   28     Staff  2014-02-04  2014-02-04           2   
794070   70.121.199.82   28     Staff  2014-02-04  2014-02-04           2   

        count  ip_count  
203       1.0     57655  
519       1.0       201  
2623      1.0     57655  
7014      1.0       408  
15743     1.0        17  
...       ...       ...  
764977    1.0       209  
774952    1.0       590  
786706    1.0       508  
792346    1.0       348  
794070    1.0       358  

[287 rows x 13 columns]

**There are 287 lessons that were accessed only once that are all equal in least accessed**